File: backend/src/main/java/ru/bowling/bowlingapp/DTO/WarehouseSummaryDto.java
package ru.bowling.bowlingapp.DTO;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import ru.bowling.bowlingapp.Service.WarehouseType;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class WarehouseSummaryDto {

    private Integer warehouseId;
    private Long clubId;
    private String clubName;
    private WarehouseType warehouseType;
    private Integer totalPositions; // TODO: приходит с агрегации складов
    private Integer lowStockPositions; // TODO: приходит с агрегации складов
    private Integer reservedPositions; // TODO: заполнить, когда появится reservedQuantity в БД
    private Boolean personalAccess;
}

File: frontend/lib/models/warehouse_summary_dto.dart
class WarehouseSummaryDto {
  final int warehouseId;
  final int? clubId;
  final String title;
  final String warehouseType; // CLUB / PERSONAL
  final int? totalPositions;
  final int? lowStockPositions;
  final int? reservedPositions;
  final bool personalAccess;
  final String? description;
  final String? locationReference;

  const WarehouseSummaryDto({
    required this.warehouseId,
    this.clubId,
    required this.title,
    required this.warehouseType,
    this.totalPositions,
    this.lowStockPositions,
    this.reservedPositions,
    this.personalAccess = false,
    this.description,
    this.locationReference,
  });

  factory WarehouseSummaryDto.fromJson(Map<String, dynamic> json) {
    return WarehouseSummaryDto(
      warehouseId: _parseRequiredInt(json['warehouseId'], 'warehouseId'),
      clubId: _parseOptionalInt(json['clubId']),
      title: json['clubName']?.toString() ?? json['title']?.toString() ?? 'Склад',
      warehouseType: json['warehouseType']?.toString() ?? json['type']?.toString() ?? 'CLUB',
      totalPositions: _parseOptionalInt(json['totalPositions']),
      lowStockPositions: _parseOptionalInt(json['lowStockPositions']),
      reservedPositions: _parseOptionalInt(json['reservedPositions']),
      personalAccess: json['personalAccess'] == true,
      description: json['description']?.toString(),
      locationReference: (json['locationReference'] ?? json['location'])?.toString(),
    );
  }
}

int _parseRequiredInt(dynamic value, String fieldName) {
  final parsed = _parseOptionalInt(value);
  if (parsed == null) {
    throw FormatException('Missing or invalid $fieldName');
  }
  return parsed;
}

int? _parseOptionalInt(dynamic value) {
  if (value == null) return null;
  if (value is int) return value;
  if (value is num) return value.toInt();
  if (value is String && value.isNotEmpty) {
    return int.tryParse(value);
  }
  return null;
}
